evm/src/arithmetic/add.rs

```diff
@@ -0,0 +1,206 @@
+    // 2^-LIMB_BITS in the base field.
+    let overflow_inv = F::inverse_2exp(LIMB_BITS);
+
+    let mut cy = builder.zero_extension();
+    for (a, b) in larger.zip(smaller) {
+        // t0 = cy + a
+        let t0 = builder.add_extension(cy, a);
+        // t  = t0 - b
+        let t = builder.sub_extension(t0, b);
+        // t1 = overflow - t
+        let t1 = builder.sub_extension(overflow, t);
+        // t2 = t * t1
+        let t2 = builder.mul_extension(t, t1);
+
+        let filtered_limb_constraint = builder.mul_extension(is_op, t2);
+        yield_constr.constraint(builder, filtered_limb_constraint);
+
+        cy = builder.mul_const_extension(overflow_inv, t);
+    }
+}
+
+pub fn generate<F: RichField>(lv: &mut [F; NUM_ARITH_COLUMNS]) {
+    let input0_limbs = ADD_INPUT_0.map(|c| lv[c].to_canonical_u64());
+    let input1_limbs = ADD_INPUT_1.map(|c| lv[c].to_canonical_u64());
+
+    // Input and output have 16-bit limbs
+    let mut output_limbs = [0u64; N_LIMBS];
+
+    const MASK: u64 = (1u64 << LIMB_BITS) - 1u64;
+    let mut cy = 0u64;
+    for (i, (&a, &b)) in input0_limbs.iter().zip(input1_limbs.iter()).enumerate() {
+        let s = a + b + cy;
+        cy = s >> LIMB_BITS;
+        debug_assert!(cy <= 1u64, "input limbs were larger than 16 bits");
+        output_limbs[i] = s & MASK;
+    }
+    // last carry is dropped because this is addition modulo 2^256.
+
+    for (&c, &output_limb) in ADD_OUTPUT.iter().zip(output_limbs.iter()) {
+        lv[c] = F::from_canonical_u64(output_limb);
+    }
+}
+
+pub fn eval_packed_generic<P: PackedField>(
+    lv: &[P; NUM_ARITH_COLUMNS],
+    yield_constr: &mut ConstraintConsumer<P>,
+) {
+    range_check_error!(ADD_INPUT_0, 16);
+    range_check_error!(ADD_INPUT_1, 16);
+    range_check_error!(ADD_OUTPUT, 16);
+
+    let is_add = lv[IS_ADD];
+    let input0_limbs = ADD_INPUT_0.iter().map(|&c| lv[c]);
+    let input1_limbs = ADD_INPUT_1.iter().map(|&c| lv[c]);
+    let output_limbs = ADD_OUTPUT.iter().map(|&c| lv[c]);
+
+    // This computed output is not yet reduced; i.e. some limbs may be
+    // more than 16 bits.
+    let output_computed = input0_limbs.zip(input1_limbs).map(|(a, b)| a + b);
+
+    eval_packed_generic_are_equal(yield_constr, is_add, output_computed, output_limbs);
+}
+
+#[allow(clippy::needless_collect)]
+pub fn eval_ext_circuit<F: RichField + Extendable<D>, const D: usize>(
+    builder: &mut plonky2::plonk::circuit_builder::CircuitBuilder<F, D>,
+    lv: &[ExtensionTarget<D>; NUM_ARITH_COLUMNS],
+    yield_constr: &mut RecursiveConstraintConsumer<F, D>,
+) {
+    let is_add = lv[IS_ADD];
+    let input0_limbs = ADD_INPUT_0.iter().map(|&c| lv[c]);
+    let input1_limbs = ADD_INPUT_1.iter().map(|&c| lv[c]);
+    let output_limbs = ADD_OUTPUT.iter().map(|&c| lv[c]);
+
+    let output_computed = input0_limbs
+        .zip(input1_limbs)
+        .map(|(a, b)| builder.add_extension(a, b))
+        .collect::<Vec<ExtensionTarget<D>>>();
```

Comment:
Since `map` is lazy and the closure passed to it borrows `builder`, I can't then borrow `builder` again below on line 135. The solution is to force evaluation with `collect`. :roll_eyes: (There is a long history of problems with clippy's handling of `needless_collect`; this seems to be another such.) Anyway, I should explicitly document this state of affairs.